<font size=5.5><b> Задача мэтчинга товаров </b></font>

**Цель работы:** Разработать алгоритм , позволяющий каждому товару из валидационной выборки сопоставить наиболее близкие товары из базовой выборки.<br>

# Загрузка данных 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import faiss
from sklearn.preprocessing import StandardScaler, RobustScaler

/tmp/ipykernel_6017/3035118191.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
df_base = pd.read_csv('base.csv', index_col=0) 
df_train = pd.read_csv('train.csv', index_col=0) 
df_test = pd.read_csv('validation.csv', index_col=0)
y_test = pd.read_csv('validation_answer.csv', index_col=0)

# Анализ данных

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
df_train.head()

0          1           2          3           4           5  \
Id                                                                             
0-query -53.882748  17.971436  -42.117104 -183.93668  187.517490  -87.144930   
1-query -87.776370   6.806268  -32.054546 -177.26039  120.803330  -83.810590   
2-query -49.979565   3.841486 -116.118590 -180.40198  190.128430  -50.837620   
3-query -47.810562   9.086598 -115.401695 -121.01136   94.652840 -109.255410   
4-query -79.632126  14.442886  -58.903397 -147.05254   57.127068  -16.239529   

                  6          7           8           9         10          11  \
Id                                                                              
0-query -347.360606  38.307602  109.085560   30.413513  -88.08269  -52.695540   
1-query  -94.572749 -78.433090  124.915900  140.331070 -177.60580  -84.995514   
2-query   26.943937 -30.447489  125.771164  211.607820  -86.34656  -35.666546   
3-query -775.150134  79.186520  124.003100  242.650650 -146.51707 -159.469850   
4-query -321.317964  45.984676  125.941284  103.392670 -107.15302   -8.800034   

                12         13          14         15         16          17  \
Id                                                                            
0-query -27.692442   4.872923  198.348000 -30.075249  -3.803569 -135.810610   
1-query  42.810810 -57.256332   96.792534 -19.261467   0.739535   50.619213   
2-query  16.395317 -80.802850  137.908650 -23.532760 -47.256584  -16.650242   
3-query -13.844755  -6.113928  118.939255 -44.585907   9.559358   14.435648   
4-query -50.977800  29.457338  143.389310   5.614824 -45.274760    9.643625   

                18         19        20           21         22          23  \
Id                                                                            
0-query -161.84137 -78.512180 -94.42894   898.436927 -70.140520   78.420360   
1-query -155.26703 -78.659430 -92.76149   353.157741 -34.744545   82.487110   
2-query -194.50568 -78.372925 -69.32448  1507.231274 -52.500970  -34.165775   
3-query -156.90683 -78.789320 -78.73709  1507.231274  19.957405   34.834290   
4-query  -77.55463 -79.066610 -77.92646  1507.231274  16.612400  116.284290   

                 24          25          26         27          28         29  \
Id                                                                              
0-query  108.032776  813.770071  -74.790880  12.610422 -183.821840  149.71584   
1-query  -28.450592  813.770071 -137.529630  26.595627 -136.783450  153.35791   
2-query   52.958652  813.770071  -18.021725  20.951107  -50.321780  158.76062   
3-query   -8.820732  813.770071 -125.606800  17.584084  -58.452904  141.28180   
4-query   33.754898  813.770071 -105.765335   6.523008  -19.812988  157.69392   

                30         31          32           33        34         35  \
Id                                                                            
0-query -27.042316 -133.21217  106.420746  -303.939233  48.73079  58.185707   
1-query  48.810093 -115.92215   87.464220  -222.286354  25.12415  91.887140   
2-query   0.178065 -183.06967   99.053570 -1018.469545 -51.80112  97.766770   
3-query -54.959310 -136.98854   63.880493 -1018.469545  89.22893  65.919960   
4-query -20.604088 -146.59128   78.849570  -780.449185  87.56077  73.036660   

                36         37          38         39         40        41  \
Id                                                                          
0-query  24.250950 -151.22410 -104.282265 -34.492810 -12.587054  2.622891   
1-query -30.636870 -136.59314 -140.500120 -43.449757  -7.226884  8.265747   
2-query -10.865850 -144.42316 -133.819490 -78.902300 -17.200352  4.467452   
3-query -24.078644 -152.33410  -91.199380 -28.225390  -4.767386  0.158236   
4-query  16.891030 -144.65790 -116.122150 -19.353254  -7.709266 -5.394988   

                 42         43         44          45          46         47  \
Id                                    

In [5]:
df_train.describe()

0              1              2              3  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -85.328679       7.664345     -43.667046    -146.118630   
std        25.803845       4.955651      39.111064      20.434841   
min      -186.280270     -11.560507    -224.896060    -223.307220   
25%      -103.309147       4.313784     -69.494487    -160.028520   
50%       -85.273695       7.652854     -42.830246    -146.067445   
75%       -67.330810      10.981191     -17.713690    -132.277745   
max        14.585236      28.917845     128.108460     -60.751625   

                   4              5              6              7  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      111.770592     -73.178792    -440.615118      21.231588   
std        47.700958      28.718853     279.017550      65.985214   
min       -93.272020    -184.962450    -791.469482    -278.582820   
25%        79.474322     -92.834080    -738.536868     -22.131764   
50%       112.260100     -73.364420    -511.837758      22.278989   
75%       143.760830     -53.699612    -201.116578      65.899595   
max       301.363600      51.849710     109.625188     288.171840   

                   8              9             10             11  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      123.067891     126.120201    -141.168584     -69.228444   
std         6.492081      64.912752      41.607785      51.258647   
min        94.831150    -142.156950    -330.570600    -352.405200   
25%       118.674642      82.545028    -168.599020    -103.048815   
50%       123.087540     125.963340    -140.461440     -68.753325   
75%       127.541802     169.701005    -113.136535     -34.668559   
max       149.356200     392.827150      37.017212     152.500000   

                  12             13             14             15  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -42.559541       6.506719     149.077582     -23.942253   
std        40.819544      37.593800      34.167843      13.024184   
min      -214.742460    -145.860280       8.818390     -82.645004   
25%       -69.801705     -18.815189     126.369124     -32.379752   
50%       -41.919796       6.739488     149.705675     -23.707227   
75%       -14.631057      32.074534     172.627318     -15.438074   
max       116.210100     171.126970     279.152560      35.552612   

                  16             17             18             19  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       -2.219093     -12.964439    -138.714610     -79.159795   
std        20.978943      55.210920      48.737245       0.775560   
min       -91.657020    -247.589280    -326.648740     -82.637184   
25%       -16.455603     -49.612492    -171.661355     -79.656816   
50%        -2.339008     -12.189709    -139.009150     -79.137460   
75%        11.702535      24.464662    -106.746862     -78.638890   
max        94.324580     215.488050      79.245160     -75.835640   

                  20             21             22             23  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -97.474574    1297.269578      -3.600685      55.908014   
std        28.382144     375.123794      50.063680      35.617513   
min      -228.808260     136.873137    -211.971800     -87.323780   
25%      -115.407545    1248.126198     -37.047511      31.436791   
50%       -96.862695    1507.231274      -3.608930      55.273705   
75%       -78.583615    1507.231274      29.473402      79.557540   
max        26.484070    1557.383334     193.884310     207.021060   

                  24             25             26             27  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       23.470181     742.788817     -64.970553      16.368072   
std        40.802653     191.052074      63.160650       4.828575   
min      -1

In [6]:
df_test.head()

0          1          2          3          4  \
Id                                                                    
100000-query -57.372734   3.597752 -13.213642 -125.92679  110.74594   
100001-query -53.758705  12.790300 -43.268543 -134.41762  114.44991   
100002-query -64.175095  -3.980927  -7.679249 -170.16093   96.44616   
100003-query -99.286860  16.123936   9.837166 -148.06044   83.69708   
100004-query -79.532920  -0.364173 -16.027431 -170.88495  165.45392   

                       5           6           7           8           9  \
Id                                                                         
100000-query  -81.279594 -461.003172  139.815720  112.880980   75.215750   
100001-query  -90.520130 -759.626065   63.995087  127.117905   53.128998   
100002-query  -62.377740 -759.626065   87.477554  131.270110  168.920320   
100003-query -133.729720   58.576403  -19.046660  115.042404   75.206730   
100004-query  -28.291668   33.931936   34.411217  128.903980  102.086914   

                     10          11         12         13         14  \
Id                                                                     
100000-query -131.89280 -140.968570 -57.987164 -22.868887  150.89552   
100001-query -153.71725  -63.951330 -52.369495 -33.390945  148.61950   
100002-query -220.30954  -31.378445  -8.788761   2.285323  133.26611   
100003-query -114.27196  -71.406456 -65.349320  24.377070   50.46730   
100004-query  -76.21417  -26.393860  34.423640  50.938890  157.68318   

                     15         16         17         18         19  \
Id                                                                    
100000-query   7.965574  17.622066 -34.868217 -216.13855 -80.908730   
100001-query -22.483830  15.164185 -56.202000 -153.61438 -79.831825   
100002-query -41.309080  14.305538 -18.231812 -205.53370 -78.160310   
100003-query -14.721335  15.069309 -46.682995 -176.60437 -78.690700   
100004-query -23.786497 -33.175415  -0.592607 -193.31854 -79.651030   

                      20           21         22          23         24  \
Id                                                                        
100000-query  -52.579520   263.363136  56.266876   66.924710  21.609911   
100001-query -101.055480  1203.537156  81.597130  101.018654  56.783424   
100002-query  -96.607670  1507.231274  -5.964200   34.937443 -56.086887   
100003-query -139.227450   325.547112   3.632292   74.929504  -4.802103   
100004-query  -91.889786  1358.481072  44.027733  121.527210  46.183000   

                      25          26         27         28         29  \
Id                                                                      
100000-query  813.770071  -32.782940  20.794031 -79.779076  156.30708   
100001-query   92.209628 -126.860340  10.382887 -38.523360  165.38391   
100002-query  813.770071  -13.200474  18.966661 -35.110190  151.36850   
100003-query  813.770071  -52.982597  15.644381 -54.087467  151.30914   
100004-query  433.623103  -82.233200  21.068508 -32.940117  149.26895   

                     30          31          32          33          34  \
Id                                                                        
100000-query -42.831330  -71.723335   83.283660 -304.174382    1.609402   
100001-query -77.840485 -169.538680  103.483240 -915.735701   16.109938   
100002-query -17.490252 -145.884300   15.533379 -655.395514   39.412827   
100003-query  21.088570 -134.507890   65.118960 -529.295053  131.565520   
100004-query   0.404718  -97.674530   81.719990 -825.644804    9.397169   

                     35         36         37          38         39  \
Id                                                                     
100000-query  55.834587 -29.474255 -139.16277 -126.038350 -62.643830   
100001-query  14.669937 -38.707085 -149.53838 -138.792920 -36.076176   
100002-query  62.554955   9.924992 -143.93462 -123.107796 -37.032475   
100003-query  67.642700 -22.884491 -145.90652  -86.917330 -11.86

In [7]:
df_test.describe()

0              1              2              3  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -85.302233       7.669724     -43.842474    -146.119797   
std        25.777321       4.956990      39.138775      20.495541   
min      -190.353330     -11.109877    -217.538420    -220.050890   
25%      -103.231790       4.292425     -69.680916    -160.210287   
50%       -85.296745       7.657888     -43.230835    -146.080365   
75%       -67.422104      10.993900     -17.765821    -132.119278   
max        14.427986      27.409784     134.859800     -57.381890   

                   4              5              6              7  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      111.635071     -73.273042    -441.396759      21.594242   
std        47.751576      28.515740     279.242585      66.478048   
min       -81.198990    -176.705030    -791.460243    -265.608430   
25%        79.101920     -92.789803    -740.620025     -21.804730   
50%       111.959330     -73.579283    -513.922978      22.138012   
75%       143.903735     -53.861139    -202.572440      66.097697   
max       302.065370      50.255325     109.632035     279.468900   

                   8              9             10             11  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      123.077725     126.001062    -141.215384     -69.340237   
std         6.504931      64.981697      41.568432      51.483403   
min        96.218350    -135.667300    -321.544250    -363.493320   
25%       118.657689      82.311205    -168.562772    -103.338022   
50%       123.062568     126.083595    -140.410225     -68.618967   
75%       127.538967     169.680500    -113.316906     -34.844935   
max       151.822560     392.827150      34.505280     170.344020   

                  12             13             14             15  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -42.626768       6.272022     149.006081     -23.961728   
std        40.858017      37.845939      34.031665      13.014670   
min      -204.299130    -144.100740     -14.816055     -78.597450   
25%       -70.147153     -19.089412     126.442184     -32.356516   
50%       -42.192298       6.420291     149.639465     -23.758115   
75%       -14.724666      31.930171     172.417195     -15.463551   
max       120.531970     177.762480     277.417850      36.471610   

                  16             17             18             19  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       -2.009594     -13.151269    -138.540600     -79.162680   
std        20.993654      55.340411      48.790143       0.777431   
min       -86.843280    -272.565600    -329.860170     -82.665300   
25%       -16.392179     -49.942476    -171.695623     -79.665475   
50%        -2.075724     -12.401822    -139.000720     -79.139723   
75%        11.971967      24.229030    -106.607981     -78.636325   
max        92.661640     197.926830      82.683810     -76.177810   

                  20             21             22             23  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -97.356361    1301.395163      -3.439855      56.097825   
std        28.503558     371.543911      50.291410      35.599512   
min      -231.376650     136.846377    -208.900850     -84.224530   
25%      -115.390365    1248.126198     -36.891637      31.741733   
50%       -96.693777    1507.231274      -3.386905      55.461280   
75%       -78.378098    1507.231274      29.878309      79.815228   
max         9.297035    1557.293478     194.321810     203.892100   

                  24             25             26             27  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       23.357262     741.156981     -64.986388      16.367853   
std        40.849607     193.199671      63.274623       4.845841   
min      -1

В данных много признаков, все численные. Сказать конкретную информацию по ним трудно. Все значения float.

# Базовое решения задачи 

In [8]:
dims = df_base.shape[1]
n_cells = 1
# how many cells
quantizer = faiss.IndexFlatL2(dims)
idx_l2 = faiss.IndexIVFFlat(quantizer, dims, n_cells)

In [9]:
idx_l2.train(np.ascontiguousarray(df_base.values[:50000, :]).astype('float32'))

idx_l2.add(np.ascontiguousarray(df_base.values).astype('float32'))

In [10]:
base_index = {k: v for k, v in enumerate(df_base.index.to_list())}

#del df_base

In [11]:
targets = df_train["Target"]
df = df_train.drop("Target", axis=1)

In [12]:
r, idx = idx_l2.search(np.ascontiguousarray(df.values).astype('float32'), 5)

In [13]:
acc = 0
for target, el in zip(targets.values.tolist(), idx.tolist()):
    acc += int(target in [base_index[r] for r in el])

print(100 * acc / len(idx))

13.804


Качество целевой метрики на сырых данных составило 13.8 , попробуем что-нибудь придумать еще. 

# Нормализация признаков

Попробуем сделать предобработку численных признаков .

In [14]:
y_train = df_train['Target'] # целевой признак обучающей выборки
X_train = df_train.drop('Target', axis=1) # обучающая выборка

In [15]:
def acc_5(targets, indices, base_indices):
   
    acc = 0 
    for target, el in zip(targets.values.tolist(), indices.tolist()):
        acc += int(target in [base_indices[r] for r in el])

    return 100 * acc / len(indices)

In [16]:
num_cols = X_train.select_dtypes(include='number').columns # численные признаки

In [17]:
# проверка
num_cols

Index(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
       '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24',
       '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36',
       '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48',
       '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60',
       '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71'],
      dtype='object')

Выполним масштабирование численных признаков в полученных датафреймах признаков с помощью `StandardScaler`.

In [44]:
# удалим X_train
del X_train

In [19]:
%%time
scaler = StandardScaler()

# обучим базовую выборку с отмасштабированными численными признаками
X_base = pd.DataFrame(scaler.fit_transform(df_base[num_cols]), index=df_base.index, columns=num_cols)

CPU times: user 2.11 s, sys: 902 ms, total: 3.01 s
Wall time: 3.01 s


In [20]:
X_train = pd.DataFrame(scaler.transform(df_train[num_cols]), index=df_train.index, columns=num_cols)
X_test = pd.DataFrame(scaler.transform(df_test[num_cols]), index=df_test.index, columns=num_cols)

In [21]:
dims = X_base.shape[1] 
n_cells = 50 
quantizer = faiss.IndexFlatL2(dims)  
idx_l2 = faiss.IndexIVFFlat(quantizer, dims, n_cells) 

In [22]:
idx_l2.train(np.ascontiguousarray(X_base.values[:50000, :]).astype('float32'))

In [23]:
%%time
idx_l2.add(np.ascontiguousarray(X_base.values).astype('float32'))

CPU times: user 16.9 s, sys: 1.37 s, total: 18.3 s
Wall time: 2.36 s


Выполним поиск

In [24]:
%%time
K_NEIGHBORS = 5
distx, idx = idx_l2.search(np.ascontiguousarray(X_train.values).astype('float32'), K_NEIGHBORS)

CPU times: user 21min 51s, sys: 1.74 s, total: 21min 52s
Wall time: 1min 51s


In [25]:
%%time
print(acc_5(y_train, idx, base_index), '%')

54.981 %
CPU times: user 379 ms, sys: 7.98 ms, total: 387 ms
Wall time: 268 ms


Видим, что значение метрики существенно возрасло: 55%

# Сжатие с потерями

Попробуем улучшить качество решаемой задачи за счет исследования различных опций алгоритма FAISS, Product Quantization (PQ).

In [26]:
dims = X_base.shape[1] 
n_cells = 50 

m = 8  
bits = 8 

quantizer = faiss.IndexFlatL2(dims) # 
idx_l2 = faiss.IndexIVFPQ(quantizer, dims, n_cells, m, bits) 

In [27]:
idx_l2.train(np.ascontiguousarray(X_base.values[:50000, :]).astype('float32'))

In [28]:
%%time
idx_l2.add(np.ascontiguousarray(X_base.values).astype('float32'))

CPU times: user 1min 8s, sys: 395 ms, total: 1min 8s
Wall time: 6.58 s


In [29]:
%%time
distx, idx = idx_l2.search(np.ascontiguousarray(X_train.values).astype('float32'), K_NEIGHBORS)

CPU times: user 1min 25s, sys: 43.9 ms, total: 1min 25s
Wall time: 7.26 s


In [30]:
%%time
print(acc_5(y_train, idx, base_index), '%')

49.176 %
CPU times: user 250 ms, sys: 7.99 ms, total: 258 ms
Wall time: 257 ms


Как видим Product Quantization не позволил повысить качество модели: 49%

## Общий вывод

В результате проеделанной работы с помошью библиотеки faiss был разработан алгорит матчинга. Алгоритм был обучен на сырых, предобработанных данных и с применением Product Quantization. Результаты по accuracy@5 приведены в таблице.
  
  
| Подход в обучении | Accuracy 5 |
|----------|---------- |
| Сырые данные            | 13.8% |
| Нормализация            | 55%   |
| Product Quantization    | 49%   |

